Updated Patient View

eknelson17 opened the pull request:
- Follows the updated design Jeff put out
- Patient Demographic information is loaded properly aside from the language (shows code) and provider (hardcoded)
- Patient History is loaded and displayed
- Left sidebar is still hardcoded

eknelson17 committed:
Adding initial remake of the patient view

- Added in Jeff's HTML and CSS - it doesn't necessarily meet all of our
  standards. (CSS needs major clean up after things are working)
- wired up the header (mostly) but had to skip provider for now and
  language is just a three letter code.
- Added font awesome to get the icons to appear on the page. Glyphicons
  will need to be replaced on the other screens eventually so we don't
  have 2 icon fonts required.
- Adding a partial test that I started work on a week or so ago. Needs
  the server to be mocked out in order to work but it is passing atm.
- Fixed tests broken by adding the new view
- Have all the "events" (procedures, conditions, encounters, and
  results) compiled into an array, but can't convert to Collection
- Inserted super basic template into show.hbs just to try to get
  something to display

eknelson17 committed:
Overhauled the patient models

- Replaced ugly copy pasta code with more modular solution
- Broke out entry types into their own models
- Patient view properly displays entries in reverse chronological order
  with the correct type. Still missing details and correct icon.

eknelson17 committed:
Overhauled the patient models. Entries now correct.

- Replaced ugly copy pasta code with more modular solution
- Broke out entry types into their own models
- Patient view properly displays entries in reverse chronological order
  with the correct type. Still missing details and correct icon.

eknelson17 committed:
Entries have icons and are displayed properly!

- Added icon property on models that is passed up to the template.
- Removed deprecated partials
- Took out all the details in order to give just a high level view
- Added a placeholder for if there is no description for the entry

eknelson17 committed:
Entries on Patient Page now have icons!

- Added icon property on models that is passed up to the template.

eknelson17 committed:
Patient Entries are listed correctly

...and are Jeff approved.

- Removing no longer needed partials
- Took out all the details that Jeff told me not to use in order to give
  just a high level view
- Added a placeholder for if there is no description for the entry

eknelson17 committed:
Cleaned up Patient View Structure and Styles

- Removed panel classes and replaced them more appropriate and semantic
  classes
- Removed excess classes and styles
- Edited existing styles to look a little better
- Reformatted styles to have a clean structure
- Added > to the left side of the patient entries
- Adjusted spacing of icons and boxes to accommodate the arrows
- Changed the "No Details..." to use bootstrap's 'text-muted' class

eknelson17 committed:
Patient View Styles are complete and cleaned up

- Removed panel classes and replaced them more appropriate and semantic
  classes
- Removed excess classes and styles
- Edited existing styles to look a little better
- Reformatted styles to have a clean structure
- Added > to the left side of the patient entries
- Adjusted spacing of icons and boxes to accommodate the arrows
- Changed the "No Details..." to use bootstrap's 'text-muted' class
- Shortened the > to be less severe
- Switched Measures to be buttons so they are ready to have a "filter"
  event tied to them in the future

carpeliam commented on app/assets/javascripts/models/patient.js.coffee:
```diff
@@ -1,4 +1,75 @@
+  parse: (attrs) ->
+    attrs = $.extend {}, attrs
+    attrs.entries = new Thorax.Collections.Entries
+    attrs.entries.comparator = (e) ->  -1 * e.get('start_time')
```
Does it make sense to define the comparator on the model itself below, instead of defining it only on this instance?

carpeliam commented on app/assets/javascripts/models/patient.js.coffee:
```diff
@@ -1,4 +1,75 @@
+    attrs = $.extend {}, attrs
+    attrs.entries = new Thorax.Collections.Entries
+    attrs.entries.comparator = (e) ->  -1 * e.get('start_time')
+    for type in ['allergies', 'conditions', 'encounters', 'immunizations', 
```
Both the Patient model and the Entries model are being held responsible for knowing about the different types; this might be better if you had an array on the Entries class defining the list of types, and then referenced it here via `for type in Thorax.Collections.Entries::types`. That way, only one class would be responsible for knowing the different types.

carpeliam commented on app/assets/javascripts/views/patient_view.js.coffee:
```diff
@@ -1,7 +1,49 @@
   # Helper function for date/time conversion
-  format_time = (time) -> moment(time).format('MM/DD/YYYY') if time
\ No newline at end of file
+  format_time = (time) -> moment(time).format('DD MMM YYYY') if time
```
As `format_time` isn't an attribute, should this be camel cased?

eknelson17 replied:
This was changed to formatTime on my local files, but for some reason it isn't showing up on GitHub even though git is telling me I have no changes.
![screen shot 2013-12-13 at 9 24 22 am](https://f.cloud.github.com/assets/1696392/1742814/5878e8fa-6402-11e3-9f8f-3dbc2b24b670.png)
![screen shot 2013-12-13 at 9 24 43 am](https://f.cloud.github.com/assets/1696392/1742815/5b73839e-6402-11e3-8ade-2c3dfbf3b093.png)

carpeliam commented on app/assets/stylesheets/styles.css.less:
```diff
@@ -153,4 +166,160 @@
+
+// patient view
+// TODO clean up the following code
+.left-sidebar {
```
Hm, this is a really nitpicky thing but I'm tempted to rename this to `sidebar` instead of `left-sidebar`; otherwise we're kind of mixing semantics with placement, and we wouldn't want to change all references if we suddenly wanted the sidebar to be on the right. Not a big deal, so if you don't agree, that's fine. Also, should this be scoped to the page that it's being applied to?

carpeliam commented on spec/javascripts/views/patient_view.spec.js.coffee:
```diff
@@ -8,11 +8,9 @@ describe 'PatientView', ->
 
   it 'formats the effective time correctly', ->
-    expect(@patientView.formatted_effective_time()).toEqual("01/16/1970")
+    expect(@patientView.context().effective_time).toEqual("16 Jan 1970")
```
Hmm.. I'd favor inspecting what's rendered on the page (similar to the above code, with `toContainText`) instead of inspecting the context directly; what's in the context isn't necessarily what the user sees.

carpeliam commented on vendor/assets/components/jQuery-Knob/LICENSE:
```diff
@@ -0,0 +1,20 @@
+The MIT License (MIT)
```
Including this file will actually break us in production :/ Sprockets has this issue where it can't process files without extensions. I've been renaming files like this to `LICENSE.txt`, but when you run bower install, it recreates files that it thinks were deleted. Kind of a pain :/ but we can't include this file.

carpeliam commented:
If we're adding something to our bower components, we should probably add it to our bower.json file as well- running bower install with the **-S** flag will do this for you.

carpeliam commented on spec/javascripts/fixtures/json/measure.json:
```diff
@@ -0,0 +1,24 @@
+{
```
This file doesn't appear to be referenced.

eknelson17 committed:
Corrected issues based on comments on pull request

- Changed all references to format_time to formatTime since it isn't an
  attribute
- Changed all references to left-sidebar to sidebar to be more semantic
- Remove unused fixture measure.json
- Deleted redundant LICENSE file for jQuery Knob
- Created a list of entry types on the Entries Collection that is used
  by Patient when creating the list of entries
- Moved the comparator for Entries from being on the instance to being
  on the class
- Added font-awesome to bower.json file
- Switched patient view tests to look at HTML instead of the context
  when checking the dates

eknelson17 committed:
Adding missed changes

- left-sidebar is now actually sidebar
- format_time is now actually formatTime

eknelson17 committed:
Fixed the dates

- Dates for patient data are now multiplied by 1000 to convert to
  milliseconds in order to use JS dates

eknelson17 committed:
Fixed Jasmine Test

- Fixed the Jasmine test for the patient view so it correctly tests the
  effective date and birthdate

eedrummer commented:
Rebased and merged
